src/Native/Unix/System.Native/pal_signal.cpp

```diff
@@ -0,0 +1,262 @@
+#include "pal_io.h"
+#include "pal_utilities.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+static struct sigaction g_origSigIntHandler, g_origSigQuitHandler; // saved signal handlers for ctrl handling
+static struct sigaction g_origSigContHandler, g_origSigChldHandler; // saved signal handlers for reinitialization
+static volatile CtrlCallback g_ctrlCallback = nullptr; // Callback invoked for SIGINT/SIGQUIT
+static int g_signalPipe[2] = {-1, -1}; // Pipe used between signal handler and worker
+
+static struct sigaction* OrigActionFor(int sig)
+{
+    switch (sig)
+    {
+        case SIGINT:  return &g_origSigIntHandler;
+        case SIGQUIT: return &g_origSigQuitHandler;
+        case SIGCONT: return &g_origSigContHandler;
+        case SIGCHLD: return &g_origSigChldHandler;
+    }
+
+    assert(false);
+    return nullptr;
+}
+
+static void SignalHandler(int sig, siginfo_t* siginfo, void* context)
+{
+    if (sig == SIGCONT || sig == SIGCHLD)
+    {
+        // SIGCONT will be sent when we're resumed after suspension, at which point
+        // we need to set the terminal back up.  Similarly, SIGCHLD will be sent after
+        // a child process completes, and that child could have left things in a bad state,
+        // so we similarly need to reinitialize.
+        ReinitializeConsole();
+    }
+
+    // Signal handler for signals where we want our background thread to do the real processing.
+    // It simply writes the signal code to a pipe that's read by the thread.
+    if (sig == SIGQUIT || sig == SIGINT || sig == SIGCHLD)
+    {
+        // Write the signal code to the pipe
+        uint8_t signalCodeByte = static_cast<uint8_t>(sig);
+        ssize_t writtenBytes;
+        while (CheckInterrupted(writtenBytes = write(g_signalPipe[1], &signalCodeByte, 1)));
+
+        if (writtenBytes != 1)
+        {
+            abort(); // fatal error
+        }
+    }
+
+    // Delegate to any saved handler we may have
+    if (sig == SIGCONT)
+    {
+        struct sigaction* origHandler = OrigActionFor(sig);
+        if (origHandler->sa_sigaction != nullptr &&
+            reinterpret_cast<void*>(origHandler->sa_sigaction) != reinterpret_cast<void*>(SIG_DFL) &&
+            reinterpret_cast<void*>(origHandler->sa_sigaction) != reinterpret_cast<void*>(SIG_IGN))
+        {
+            origHandler->sa_sigaction(sig, siginfo, context);
+        }
+    }
+}
+
+__attribute__((unused)) static void ResumeSIGCHLD()
+{
+    struct sigaction* origHandler = OrigActionFor(SIGCHLD);
+
+    if (reinterpret_cast<void*>(origHandler->sa_sigaction) == reinterpret_cast<void*>(SIG_IGN))
+    {
+        // When the disposition is SIG_IGN, children that terminated do not become zombies.
+        pid_t pid;
+        do
+        {
+            int status;
+            while (CheckInterrupted(pid = waitpid(WAIT_ANY, &status, WNOHANG)));
```

Comment:
Why is it necessary to do waitpid if it's not to clear out zombies?

Reply:
The comment is confusing. I'll reword it.

If the original disposition is SIG_IGN, then the kernel won't generate zombies. (comment)
But since we overwrote the disposition, we do get zombies, and we need to waitpid them. (code)